test_conformance/extensions/cl_khr_command_buffer/cl_khr_command_buffer_mutable_dispatch/command_buffer_test_mutable_dispatch.cpp

```diff
@@ -0,0 +1,590 @@
+                & CL_DEVICE_MUTABLE_DISPATCH_CAPABILITIES_KHR)
+                != 0;
+
+        return !mutable_support || !extension_avaliable
+            || BasicCommandBufferTest::Skip();
+    }
+
+    cl_int init_extension_functions()
+    {
+        BasicCommandBufferTest::init_extension_functions();
+
+        cl_platform_id platform;
+        cl_int error =
+            clGetDeviceInfo(device, CL_DEVICE_PLATFORM, sizeof(cl_platform_id),
+                            &platform, nullptr);
+        test_error(error, "clGetDeviceInfo for CL_DEVICE_PLATFORM failed");
+
+        // If it is supported get the addresses of all the APIs here.
+#define GET_EXTENSION_ADDRESS(FUNC)                                            \
+    FUNC = reinterpret_cast<FUNC##_fn>(                                        \
+        clGetExtensionFunctionAddressForPlatform(platform, #FUNC));            \
+    if (FUNC == nullptr)                                                       \
+    {                                                                          \
+        log_error("ERROR: clGetExtensionFunctionAddressForPlatform failed"     \
+                  " with " #FUNC "\n");                                        \
+        return TEST_FAIL;                                                      \
+    }
+        GET_EXTENSION_ADDRESS(clGetMutableCommandInfoKHR);
+
+        return CL_SUCCESS;
+    }
+
+    clGetMutableCommandInfoKHR_fn clGetMutableCommandInfoKHR = nullptr;
+    const char* kernelString = "__kernel void empty() {}";
+};
+
+struct InfoDeviceQuery : public BasicMutableCommandBufferTest
+{
+    using BasicMutableCommandBufferTest::BasicMutableCommandBufferTest;
+
+    InfoDeviceQuery(cl_device_id device, cl_context context,
+                    cl_command_queue queue)
+        : BasicMutableCommandBufferTest(device, context, queue)
+    {}
+
+    cl_int Run() override
+    {
+        cl_mutable_dispatch_fields_khr mutable_capabilities;
+
+        cl_int error = clGetDeviceInfo(
+            device, CL_DEVICE_MUTABLE_DISPATCH_CAPABILITIES_KHR,
+            sizeof(mutable_capabilities), &mutable_capabilities, nullptr);
+        test_error(error, "clGetDeviceInfo failed");
+
+        if (!(mutable_capabilities & CL_MUTABLE_DISPATCH_ARGUMENTS_KHR))
+        {
+            log_error("Device does not support update arguments to a "
+                      "mutable-dispatch.");
+            return TEST_FAIL;
+        }
+
+        return CL_SUCCESS;
+    }
+};
+
+struct InfoBuffer : public BasicMutableCommandBufferTest
+{
+    using BasicMutableCommandBufferTest::BasicMutableCommandBufferTest;
+
+    InfoBuffer(cl_device_id device, cl_context context, cl_command_queue queue)
+        : BasicMutableCommandBufferTest(device, context, queue),
+          test_command_buffer(this)
+    {}
+
+    cl_int Run() override
+    {
+        cl_int error = clCommandNDRangeKernelKHR(
+            command_buffer, nullptr, nullptr, kernel, 1, nullptr, nullptr,
```

Comment:
It's a test for `CL_MUTABLE_COMMAND_COMMAND_BUFFER_KHR`. How come I should `global_work_size` here?

Reply:
The spec wording for `clCommandNDRangeKernelKHR` says

```quote
clCommandNDRangeKernelKHR returns CL_SUCCESS if the function is executed successfully. Otherwise, it returns the errors defined by clEnqueueNDRangeKernel except:
```

And [clEnqueueNDRangeKernel](https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_API.html#clEnqueueNDRangeKernel) has error 

```
CL_INVALID_GLOBAL_WORK_SIZE if global_work_size is NULL or ...
```

An implementation won't be able to properly create the command without this information, and there is no guarantee a user will later call update() to set it